src/passes/MemoryPacking.cpp

```diff
@@ -23,141 +23,394 @@
 
 namespace wasm {
 
-// Adding segments adds overhead, this is a rough estimate
-const Index OVERHEAD = 8;
+// A subsection of an orginal memory segment
+struct Range {
+  bool isZero;
+  size_t start;
+  size_t end;
+};
 
-struct MemoryPacking : public Pass {
-  bool modifiesBinaryenIR() override { return false; }
+// An initial segment index and a list of ranges an original segment has been
+// split into. This is enough information to replace a bulk memory operation on
+// an original segment with an equivalent sequence of operations on the new,
+// split segments.
+using Replacement = std::pair<Index, std::vector<Range>>;
+
+// Maps bulk memory operations to the replacement that must be applied to them.
+using Replacements = std::unordered_map<Expression*, Replacement>;
 
+// memory.fill: 2 byte opcode + ~2 immediate bytes + ~(3*2) byte operands
+const size_t MEMORY_FILL_SIZE = 10;
+
+// data.drop: 2 byte opcode + ~1 byte index immediate
+const size_t DATA_DROP_SIZE = 3;
+
+struct MemoryPacking : public Pass {
   void run(PassRunner* runner, Module* module) override {
     if (!module->memory.exists) {
       return;
     }
 
+    auto& segments = module->memory.segments;
+
+    // For each segment, a list of bulk memory instructions that refer to it
+    std::vector<std::vector<Expression*>> referers(segments.size());
```

Comment:
Oops, I've fallen into the [classic trap](https://en.wikipedia.org/wiki/HTTP_referer). I would prefer `referrer` over `accessor` because `data.drop` does not really access the segment.